servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/TimeoutPublisher.java

```diff
@@ -29,53 +28,55 @@
+ * <ul>
+ *     <li>An absolute timeout which requires that the publisher receive all items and complete (or error)</li>
+ *     <li>A mode which restarts the timer for each {@link Subscriber#onNext(Object) onNext} signal providing an "idle"
+ *     timeout.</li>
+ * </ul>
+ *
+ * <p/>For either timer mode the timer begins with the subscription.
+ *
+ * @param <T> Type of items
+ */
 final class TimeoutPublisher<T> extends AbstractNoHandleSubscribePublisher<T> {
     private final Publisher<T> original;
     private final io.servicetalk.concurrent.Executor timeoutExecutor;
+    /**
+     * non-negative nanoseconds until the timeout
+     */
     private final long durationNs;
-
-    TimeoutPublisher(final Publisher<T> original,
-                     final Executor executor,
-                     final Duration duration) {
-        this(original, executor, duration, executor);
-    }
-
-    TimeoutPublisher(final Publisher<T> original,
-                     final Executor executor,
-                     final long duration,
-                     final TimeUnit unit) {
-        this(original, executor, duration, unit, executor);
-    }
-
-    TimeoutPublisher(final Publisher<T> original,
-                     final Executor publisherExecutor,
-                     final Duration duration,
-                     final io.servicetalk.concurrent.Executor timeoutExecutor) {
-        this(original, publisherExecutor, duration.toNanos(), timeoutExecutor);
-    }
+    /**
+     * If true then restart the timeout at every {@link TimeoutSubscriber#onNext(Object)} otherwise start only at
+     * subscribe
+     */
+    private final boolean restartAtOnNext;
 
     TimeoutPublisher(final Publisher<T> original,
                      final Executor publisherExecutor,
                      final long duration,
                      final TimeUnit unit,
+                     final boolean restartAtOnNext,
                      final io.servicetalk.concurrent.Executor timeoutExecutor) {
-        this(original, publisherExecutor, unit.toNanos(duration), timeoutExecutor);
+        this(original, publisherExecutor, unit.toNanos(duration), restartAtOnNext, timeoutExecutor);
     }
 
-    private TimeoutPublisher(final Publisher<T> original,
+    TimeoutPublisher(final Publisher<T> original,
                              final Executor publisherExecutor,
                              final long durationNs,
+                             final boolean restartAtOnNext,
                              final io.servicetalk.concurrent.Executor timeoutExecutor) {
         super(publisherExecutor);
         this.original = requireNonNull(original);
         this.timeoutExecutor = requireNonNull(timeoutExecutor);
         // We use the duration in arithmetic below to determine the expiration time for the "next timer" below. So
         // lets cap this at 0 to simplify overflow at that time.
         this.durationNs = max(0, durationNs);
```

Comment:
It's pre-existing, but wdyt if instead of forcing it to 0 we validate and thow `IllegalArgumentException` when duration is less than 1ns? Or we can do that in a follow-up to extract the behavior change out of this PR.

Reply:
I think the forcing to zero is intended to allow sloppiness around calculated durations. Like the various java.util.concurrent timed operations <= 0 values result in no waiting. It also normalizes the timeout behaviour such that a non-positive value always just times out immediately.